DSA/4.1 Stacks/17. Minimum Or Maximum Stack Implementation/MinimumOrMaximumStack.md

```diff
@@ -0,0 +1,230 @@
+
+<br>
+
+## **Pseudo Code**
+
+<br>
+
+```txt
+class Stack
+{
+  current_size
+  vector v
+  vector maximum
+
+public:
+  Stack()
+  {
+    current_size <- 0
+  }
+
+  void Push(element)
+  {
+    if current_size == 0
+      maximum <- push_back(element)
+    else
+      maximum <- push_back(max(element, maximum[curren_size - 1]))
+    v <- push_back(element)
+    current_size = current_size + 1
+  }
+
+  void Pop()
+  {
+    if current_size > 0
+      v <- pop_back()
+      maximum <- pop_back()
+      current_size <- current_size - 1
+  }
+
+  void Print_Max_Vector()
+  {
+    for(i from 1 to current_size)
+      print maximum[i - 1]
+  }
+
+  void Print_Vector()
+  {
+    for(i from 1 to current_size - 1)
+      print v[i - 1]
+  }
+
+  int Max_Element()
+  {
+    return maximum[current_size - 1]
+  }
+};
+```
+
+<br>
+
+### **Example**
+
+<br>
+
+```txt
+Suppose we have a random set of numbers: 12 4 56 3 7 100 
+We create a Stack class object, Constructor will initialize the current_size with 0
+
+Stack :
+v : [EMPTY]
+maximum : [EMPTY]
+current_size : 0
+
+Pushing 12 :
+v : [12]
+maximum : [12]
+current_size : 1
+
+Pushing 4 :
+v : [12 4]
+maximum : [12 12]
+current_size : 2
+
+Pushing 56 :
+v : [12 4 56]
+maximum : [12 12 56]
+current_size : 3
+
+Pushing 3 :
+v : [12 4 56 3]
+maximum : [12 12 56 56]
+current_size : 4
+
+Pushing 7 :
+v : [12 4 56 3 7]
+maximum : [12 12 56 56 56]
+current_size : 5
+
+Pushing 100 :
+v : [12 4 56 3 7 100]
+maximum : [12 12 56 56 56 100]
+current_size : 6
+```
+
+<br>
+
+## **Code**
+
+<br>
+
+```C++
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+class Stack
+{
+	int current_size;
+	vector <int> v;
+	vector <int> maximum;
+public:
+	Stack()
+	{
+		current_size = 0;
+	}
+
+	void push(int element)
+	{
+		if(current_size == 0)
+			maximum.push_back(element);
+		else
+			maximum.push_back(max(element, maximum[current_size - 1]));
+		v.push_back(element);
+		
+		current_size++;
+	}
+
+	void pop()
+	{
+		if(current_size > 0)
+		{
+			v.pop_back();
+			maximum.pop_back();
+			current_size--;
+		}
+	}
+
+	void print_max_vector()
+	{
+		cout << "Max stack is :\n";
+		for(int i = 0; i < current_size; i++)
+			cout << maximum[i] << " ";
+		cout << "\n\n";
+	}
+
+	void print_stack()
+	{	
+		cout << "The stack is :\n";
+		for(int i = 0; i < current_size - 1; i++)
+			cout << v[i] << " ";
+		cout << "\n\n";
+	}
+
+	int max_element()
+	{
+		return maximum[current_size - 1];
+	}
+};
+
+void take_input()
+{
+	int q;
+	cin >> q;
+	Stack s;
+
+	while(q--)
+	{
+		string str;
+		cin >> str;
+
+		if(str == "push")
+		{
+			int num;
+			cin >> num;
+			s.push(num);
+		}
+		else if(str== "pop")
+			s.pop();
+		else if(str == "max")
+			cout << s.max_element() << "\n";
+	}
+}
+
+int main()
+{
+	take_input();
+
+	return 0;
+}
+```
+
+<br>
+
+## **Output**
+
+<br>
+
+![image](https://user-images.githubusercontent.com/34866732/142176253-b8ddc9ab-a7e5-4f71-a5a3-3994e5b21662.png)
+
+<br>
```

Comment:
Add time and space complexities as well

Reply:
@nimisha-yadav I have made all the changes that were asked by you, can you please review my PR again?